Refactor Eks Create Cluster Operator code

syedahsn opened the pull request:
The `EksCreateClusterOperator` is capable of creating a Fargate Profile or Nodegroup that is associated with the newly created cluster. The code for creating the Fargate Profile and Nodegroup is the same as the code in `EksCreateFargateProfileOperator` and `EksCreateNodegroupOperator` respectively. 
This PR refactors the code to reuse some of the code to decrease code duplication. This step also simplifies adding deferrable mode to these operators.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

syedahsn committed:
Refactor EksCreateClusterOperator to reuse code being used in multiple places

phanikumv commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -59,7 +59,54 @@
 FARGATE_FULL_NAME = "AWS Fargate profiles"
 
 
-class EksCreateClusterOperator(BaseOperator):
```
Dont we need deprecation warning here, and call the new operator if we remove this class? Otherwise existing DAGs using the `EksCreateClusterOperator` will break right?

syedahsn replied:
I'm not removing the class, just making it inherit from a different class. It shouldn't affect any existing DAGs.

vincbeck replied:
Yes Github makes it confusing but look at line `107`, `EksCreateClusterOperator` is still there

syedahsn committed:
Update create_compute method to pass tests
Add waiter params to EksCreateClusterOperator and EksCreateNodegroupOperator
Update EksCreateFargateProfileTrigger and EksDeleteFargateProfileTrigger to use more consistent waiter names
Update unit tests for triggers and operators

syedahsn committed:
Add BaseEksCreateOperator to MISSING_EXAMPLES_FOR_CLASSES because it is meant to be a parent class to some of the Eks operators

o-nikolas commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -184,12 +235,12 @@ def execute(self, context: Context):
+        self.eks_hook = EksHook(
             aws_conn_id=self.aws_conn_id,
             region_name=self.region,
         )
```
Now that you have a cool base class, you can add an __init__ to it and create the hook inside of that so that all the child classes don't need to create it each time.

syedahsn replied:
Nice, that's a good idea. Thanks!

syedahsn replied:
Made the changes

o-nikolas commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -393,51 +416,41 @@ def __init__(
         self.fargate_profile_name = fargate_profile_name
         self.create_fargate_profile_kwargs = create_fargate_profile_kwargs or {}
-        self.wait_for_completion = wait_for_completion
+        self.wait_for_completion = False if deferrable else wait_for_completion
```
Why is this only done on this operator but not the previous two?

syedahsn replied:
Because as of right now, this is the only Operator that has the deferrable property. Once deferrable mode is implemented for the other 2 operators, I'll update `wait_for_completion` to match this one.

o-nikolas replied:
Ahh, gotcha, makes sense :+1:

o-nikolas commented on airflow/providers/amazon/aws/triggers/eks.py:
```diff
@@ -33,49 +33,52 @@ class EksCreateFargateProfileTrigger(BaseTrigger):
     :param fargate_profile_name: The name of the fargate profile
-    :param poll_interval: The amount of time in seconds to wait between attempts.
-    :param max_attempts: The maximum number of attempts to be made.
+    :param waiter_delay: The amount of time in seconds to wait between attempts.
```
Why all the renaming? The PR description doesn't mention this and it seems unrelated to factoring out the create compute method?

syedahsn replied:
I realized I've been naming these parameters incorrectly for the past few Triggers, and I was hoping to fix them. Considering the `EksCreateFargateProfileTrigger` got merged very recently, I was hoping to do this without doing a  deprecation warning. If it is problematic to include it as part of this PR, I can take it out.

o-nikolas replied:
> got merged very recently, I was hoping to do this without doing a deprecation warning

This should be fine since we haven't done a provider release yet

 > If it is problematic to include it as part of this PR, I can take it out.
 
In a perfect world I think a separate PR would have been best, but it's not a huge deal, so let's keep it here for this time :+1:

o-nikolas commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -59,7 +59,52 @@
+
+    """
+
+    def create_compute(cls):
```
This is an instance method, can you s/cls/self/g in all places for this method?

syedahsn committed:
Add eks_hook to base class
Minor PR comment fixes

o-nikolas commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -184,12 +238,12 @@ def execute(self, context: Context):
+        self.eks_hook = EksHook(
             aws_conn_id=self.aws_conn_id,
             region_name=self.region,
         )
```
This can be removed now right?

syedahsn replied:
sorry I missed that. Changing it now

syedahsn committed:
remove redundant eks_hook creation

uranusjr commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -59,7 +59,55 @@
+    Base class for EKS Operators that create cluster resources. This includes
+    EksCreateClusterOperator, EksCreateNodeGroupOperator and EksCreateFargateProfileOperator.
+
+    """
```
```suggestion
    """Base class for EKS Operators that create cluster resources.

    This includes EksCreateClusterOperator, EksCreateNodeGroupOperator and
    EksCreateFargateProfileOperator.
    """
```

I’m not sure we should have the second line though. The list of classes doesn’t really add much useful information for maintenance, and is very easy to get stale.

uranusjr replied:
Also maybe a `:meta private:` would be a good idea? I don’t think we expect users to call this class directly. (Or rename this class to `_BaseEksCreateOperator`)

uranusjr commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -59,7 +59,55 @@
+
+    def __init__(self, **kwargs):
+        self.eks_hook = EksHook(aws_conn_id=self.aws_conn_id, region_name=self.region)
+        super().__init__(**kwargs)
```
Any reason the hook is instantiated before the `super()` call? I believe the majority preference is to do it after the super call. (I don’t think the style matters here, but it may introduce unnecessary confusion.)

syedahsn replied:
No particular reason. I didn't really pay attention to the order :sweat_smile:

uranusjr commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -59,7 +59,55 @@
+        self.eks_hook = EksHook(aws_conn_id=self.aws_conn_id, region_name=self.region)
+        super().__init__(**kwargs)
+
+    def create_compute(self):
```
```suggestion
    def create_compute(self) -> None:
```

And now Mypy would start to prompt you adding variable definitions on the class (which is good)

syedahsn replied:
I made this change, and it ended up having a bigger impact than I anticipated. I had to add a bunch of parameters used in the 3 child classes. I think I've got everything ready. I need to do some testing with how templating works here. Right now, `_BaseEksCreateOperator` doesn't have any templated fields, but because the child classes allow templating for some parameters, I think I have to do the same for `_BaseEksCreateOperator`. 

@uranusjr what do you think?

syedahsn committed:
Add __init__ method to _BaseEksCreateOperator to accept all parameters

syedahsn committed:
Replace _BaseEksCreateOperator with _create_compute

syedahsn committed:
Remove BaseEksCreateOperator from MISSING_EXAMPLES_FOR_CLASSES
Fix unit tests
